pkg/util/startup/retry.go

```diff
@@ -0,0 +1,147 @@
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+type startupRetryKey struct{}
+
+var startupRetryOpts = retry.Options{
+	InitialBackoff: 100 * time.Millisecond,
+	MaxBackoff:     3 * time.Second,
+	Multiplier:     2,
+}
+
+// runningStartup is a counter showing number of concurrently running server
+// startups. if it is zero, then startups are finished and we should not try
+// to run startup retry checks.
+var runningStartup atomic.Int32
+
+// RetryContext adds startup tag to allow kv access from startup path.
+func RetryContext(ctx context.Context) context.Context {
+	return context.WithValue(ctx, startupRetryKey{}, "in retry")
+}
+
+// RunWithStartupRetry synchronously runs a function with retry of circuit
+// breaker errors on startup.
+func RunWithStartupRetry(
+	ctx context.Context, opName string, f func(ctx context.Context) error,
+) error {
+	ctx = RetryContext(ctx)
+	// We retry till success or till context is cancelled.
+	var err error
+	for r := retry.Start(startupRetryOpts); r.Next(); {
+		err = f(ctx)
+		if err == nil {
+			break
+		}
+		if ctx.Err() != nil || !IsRetryableReplicaError(err) {
+			break
+		}
+		log.Infof(ctx, "failed %s during node startup, retrying (%s / %s)", opName, err, reflect.TypeOf(err))
+	}
+	return err
+}
+
+// RunWithStartupRetryEx run function returning value with startup retry.
+func RunWithStartupRetryEx[T any](
+	ctx context.Context, opName string, f func(ctx context.Context) (T, error),
+) (T, error) {
+	ctx = RetryContext(ctx)
+	// Retry indefinitely until context is cancelled.
+	var result T
+	var err error
+	for r := retry.Start(startupRetryOpts); r.Next(); {
+		result, err = f(ctx)
+		if err == nil {
+			break
+		}
+		if ctx.Err() != nil || !IsRetryableReplicaError(err) {
+			break
+		}
+		log.VEventf(ctx, 1, "failed %s during node startup, retrying %s", opName, err)
+	}
+	// This shouldn't happen as we retry indefinitely.
+	return result, err
+}
+
+// RunWithoutRetry run function with context that allows kv execution query on
+// startup. This is meant for parts of code that is only executed in tests
+// e.g. code paths that avoid asynchronous execution or creating jobs and
+// instead run eagerly during startup. This shouldn't be used to suppress
+// startup assertions as makes server vulnerable to crash loops on startup.
+func RunWithoutRetry(ctx context.Context, f func(ctx context.Context) error) error {
+	ctx = RetryContext(ctx)
+	return f(ctx)
+}
+
+func MaybeTagStartupContext(ctx context.Context) func() {
+	if !util.RaceEnabled {
+		return func() {}
+	}
+	if !inStartup() {
+		log.Fatal(ctx, "startup tag context should only be called from server startup method")
+	}
+	runningStartup.Add(1)
+	return func() {
+		runningStartup.Add(-1)
+	}
+}
+
+func AssertStartupRetry(ctx context.Context) {
```

Comment:
This seems like it could have false positives. You want to check that while starting up, all goroutines that call this have a tagged context. But when startup is "done", and there is a slight delay to actually "unregister" the startup, can't there be something in the system that doesn't have the correctly tagged ctx just because it has very little to do with the startup procedure?